src/avx2/stage1.rs

```diff
@@ -10,45 +10,76 @@ use std::mem;
 
 pub const SIMDJSON_PADDING: usize = mem::size_of::<__m256i>();
 
+unsafe fn compute_quote_mask(quote_bits: u64) -> u64 {
+    _mm_cvtsi128_si64(_mm_clmulepi64_si128(
+        _mm_set_epi64x(0, static_cast_i64!(quote_bits)),
+        _mm_set1_epi8(-1 /* 0xFF */),
+        0,
+    )) as u64
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn check_ascii(input: &SimdInput) -> bool {
+    let highbit: __m256i = _mm256_set1_epi8(static_cast_i8!(0x80u8));
+    let test_v0v1 = _mm256_testz_si256(_mm256_or_si256(input.v0, input.v1), highbit);
+
+    test_v0v1 == 1
+}
+
 #[derive(Debug)]
 struct SimdInput {
-    lo: __m256i,
-    hi: __m256i,
+    v0: __m256i,
+    v1: __m256i,
 }
 
 fn fill_input(ptr: &[u8]) -> SimdInput {
     unsafe {
         #[allow(clippy::cast_ptr_alignment)]
         SimdInput {
-            lo: _mm256_loadu_si256(ptr.as_ptr() as *const __m256i),
-            hi: _mm256_loadu_si256(ptr.as_ptr().add(32) as *const __m256i),
+            v0: _mm256_loadu_si256(ptr.as_ptr() as *const __m256i),
+            v1: _mm256_loadu_si256(ptr.as_ptr().add(32) as *const __m256i),
         }
     }
 }
 
+struct Utf8CheckingState {
```

Comment:
I am not sure if that has a performance impact but I think putting it in a struct might prevent rust from treating it as a register to be passed (and by that hurt performance) - we should verify that.

Reply:
and I presume the same thing if it's a tuple?

Reply:
To be honest I'm not sure one way or the other, I'm just suspecting that if we return a struct it gets put in the memory rather then registers but I could be totally wrong.